### PR TITLE
Wip/xz

### PR DESCRIPTION
**Pull request by nacho**
These few patches should be good enough to backport

### CODE REVIEW

**Comment by bviktor** [comment]
I dunno why exactly c:\mozilla-build\hexchat bothers you but FYI it's also used during building HexChat.

**Comment by nacho** [comment]
grrr, that commit should not have gone on the pull request, please cherry pick just the first 2 commits

**Comment by tomek** [comment]
@nacho could you tell me the reason why should we add .xz? .7z works just fine and that is how we are storing source for our bunde.

**Comment by Arnavion** [comment]
@tomek

The tar.xz change is useful because we can then stop repacaking the source for all the deps and just upload the originals. It's not correct right now - needs to support .tar.gz as well and also needs to check that extension is .tar.xz, not just .xz, but it's a good idea. @nacho Can you do this change? (Also force-push the third commit out of this branch while you're at it.) Alternatively I can do it and rebase-fixup into your commit.

Some deps might be a problem, eg ATK has two files with the same name but different case, so I'm not sure if 7z will complain, but for other deps it will be good.

**Comment by Arnavion** [comment]
Also, for the first commit, move it after the doc comment block.

```diff
 .LINK
 http://hexchat.github.io/gtk-win32/
 
 #>
 
+#Requires -Version 4.0
+
 #========================================================================================================================================================
 # Parameters begin here
```

File: build.ps1 [hunk 1]
```diff
@@ -956,8 +968,26 @@ $items.GetEnumerator() | %{
...
-		Exec $SevenZip x $item.ArchiveFile -o"$workingDirectory" -y > $null
-		"Extracted $($item.ArchiveFile.Name)"
+		switch ([System.IO.Path]::GetExtension($item.ArchiveFile)) {
+			'.xz' {
+				$filename = $item.ArchiveFile.toString()
+				$command = @"
+  "$SevenZip" x -y -so "$filename" 2> nul | "$SevenZip" x -y -si -ttar -aoa -o"$workingDirectory"
+"@
+
+				echo $command
```
Comment (Arnavion) [inline build.ps1]: "$command"

File: build.ps1 [hunk 2]
```diff
@@ -956,8 +968,26 @@ $items.GetEnumerator() | %{
...
+
+				cmd /c $command
```
Comment (Arnavion) [inline build.ps1]: `Exec cmd /c $command`

File: build.ps1 [hunk 3]
```diff
@@ -956,8 +968,26 @@ $items.GetEnumerator() | %{
...
+
+				"Extracted $($item.ArchiveFile.Name)"
```
Comment (Arnavion) [inline build.ps1]: It's common in all cases so move it after the switch block.

File: build.ps1 [hunk 4]
```diff
@@ -868,11 +870,21 @@ $items.GetEnumerator() | %{
...
+		'.xz' {
+			$filename = [System.IO.Path]::GetFileNameWithoutExtension($item.ArchiveFile)
+			$tarFile = New-Object System.IO.FileInfo "$workingDirectory\$filename"
+			$item.BuildDirectory = $(New-Object System.IO.DirectoryInfo "$workingDirectory\$($tarFile.BaseName)")
+		}
+
+		'.7z' {
+			$item.BuildDirectory = $(New-Object System.IO.DirectoryInfo "$workingDirectory\$($archiveFile.BaseName)")
+		}
+	}
```
Comment (Arnavion) [inline build.ps1]:
```powershell
default {
    throw "Unrecognized file extension $_"
}
```

File: build.ps1 [hunk 5]
```diff
@@ -868,11 +870,21 @@ $items.GetEnumerator() | %{
...
 	$item.ArchiveFile = $archiveFile
 	$item.PatchDirectory = $(New-Object System.IO.DirectoryInfo $patchDirectory)
-	$item.BuildDirectory = $(New-Object System.IO.DirectoryInfo "$workingDirectory\$($archiveFile.BaseName)")
-	$item.BuildArchiveFile = $(New-Object System.IO.FileInfo "$workingDirectory\$($archiveFile.BaseName)-$filenameArch$($archiveFile.Extension)")
 	$item.Dependencies = @($item.Dependencies | %{ $items[$_] })
 	$item.Dependents = @()
 	$item.State = ''
+
+	switch ([System.IO.Path]::GetExtension($item.ArchiveFile)) {
+		'.xz' {
```
Comment (Arnavion) [inline build.ps1]: As I mentioned in my comment on the thread, I'd rather this check explicitly for `'.tar.xz'` instead, so you can't use Path.GetExtension nor a switch block for this.

File: build.ps1 [hunk 6]
```diff
@@ -956,8 +968,26 @@ $items.GetEnumerator() | %{
...
+
+				echo $command
+
+				cmd /c $command
+
+				"Extracted $($item.ArchiveFile.Name)"
+			}
+
+			'.7z' {
+				Exec $SevenZip x $item.ArchiveFile -o"$workingDirectory" -y > $null
```
Comment (Arnavion) [inline build.ps1]: Add -aoa here too for consistency.

File: build.ps1 [hunk 7]
```diff
@@ -868,11 +870,21 @@ $items.GetEnumerator() | %{
...
 	$item.PatchDirectory = $(New-Object System.IO.DirectoryInfo $patchDirectory)
-	$item.BuildDirectory = $(New-Object System.IO.DirectoryInfo "$workingDirectory\$($archiveFile.BaseName)")
-	$item.BuildArchiveFile = $(New-Object System.IO.FileInfo "$workingDirectory\$($archiveFile.BaseName)-$filenameArch$($archiveFile.Extension)")
 	$item.Dependencies = @($item.Dependencies | %{ $items[$_] })
 	$item.Dependents = @()
 	$item.State = ''
+
+	switch ([System.IO.Path]::GetExtension($item.ArchiveFile)) {
+		'.xz' {
+			$filename = [System.IO.Path]::GetFileNameWithoutExtension($item.ArchiveFile)
```
Comment (Arnavion) [inline build.ps1]: $item.ArchiveFile is $archiveFile, which is already a FileInfo. You don't need to convert it implicitly to a string and then use Path functions on it, just use $archiveFile.BaseName for this.

File: build.ps1 [hunk 8]
```diff
@@ -868,11 +870,21 @@ $items.GetEnumerator() | %{
...
+			$tarFile = New-Object System.IO.FileInfo "$workingDirectory\$filename"
```
Comment (Arnavion) [inline build.ps1]: Here and the next line is where you *should* use Path.GetFileNameWithoutExtension

```powershell
$item.BuildDirectory = $(New-Object System.IO.DirectoryInfo "$workingDirectory\$([System.IO.Path]::GetFileNameWithoutExtension($archiveFile))")
```

**Comment by nacho** [comment]
Thanks a lot for the reviews, as you can imagine already this is the first time I touch powershell stuff :) Tomorrow I'll fix up this.

**Comment by nacho** [comment]
Just pushed the new patches. Let me know if there is anything else to do. Just one thing. I was not able to change:
cmd /c $command to Exec cmd /c $command

For some reason doing that gives me an error.

**Comment by Arnavion** [comment]
- Not very helpful if you don't tell us what the error is... You absolutely need to run it under Exec.

- You're still not handling .tar.gz

**Comment by nacho** [comment]
Here is the error:
gtk3 : cmd /c   "C:\Program Files\7-Zip\7z.exe" x -y -so "C:\mozilla-build\gtk\src\gtk+-3.14.8.tar.xz" 2> nul | "C:\Program Files\7-Zip\7z.exe" x -y -si -ttar -aoa -o"C:\mozilla-build\gtk\build\x64" exited with code 2

About .tar.gz I was thinking about adding the formats in different commits. See i.e right now I need tar.xz because of:
https://github.com/nice-software/gtk-win32/commit/fcb3c757e1162ea054ea2508fe54fa8fc9d2022b

**Comment by Arnavion** [comment]
Right, so exit code 2 means it's not a successful. The whole point of Exec is to detect non-0 error code and abort, because that means the process it ran didn't run successfully. Ergo something is wrong with the command you tried to get it to run.

**Comment by nacho** [comment]
dunno, I am not able to find the problem, if you could help would be great

**Comment by nacho** [comment]
So I decided to do it in 2 steps at the end to avoid the exit code 2. Also added the tar.gz and tar.bz2 (untested) but it should be the same as tar.xz afaik...

**Comment by Arnavion** [comment]
Initially you had this:

```powershell
$command = @"
"$SevenZip" x -y -so "$filename" 2> nul | "$SevenZip" x -y -si -ttar -aoa -o"$workingDirectory"
"@;
cmd /C $command
```

You decided to not use `Exec` because it threw an exception, as if implying that it worked even though it threw an exception. But it's obvious you didn't actually actually run the code you submitted a PR of, because the exception wasn't just for show.

```powershell
PS [12:07:52] C:\Users\Arnavion\Desktop> cmd /C $command
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
PS [12:08:01] C:\Users\Arnavion\Desktop> echo $LASTEXITCODE
1
```

The reason is that cmd.exe requires a double quote around the string passed with /C if it contains more than one command separated by & or |. So adding double quotes around it is sufficient.

```powershell
PS [12:10:05] C:\Users\Arnavion\Desktop> cmd /C "`"$command`""

<snip large output of successful extraction>

Everything is Ok

Size:       6645760
Compressed: 696064

Everything is Ok

Folders: 14
Files: 482
Size:       6262882
Compressed: 259584
PS [11:59:40] C:\Users\Arnavion\Desktop> echo $LASTEXITCODE
0
```

Furthermore, redirecting the stdout of the first 7z.exe with the pipe operator is sufficient to split the streams, so the NUL redirection of stderr isn't even needed. In fact it is harmful since if it fails for any reason, the error message will be sent to NUL.

```powershell
$command = @"
"$SevenZip" x -y -so "$filename" | "$SevenZip" x -y -si -ttar -aoa -o"$workingDirectory"
"@;
Exec cmd /C "`"$command`""
```

---

Note that the error code returned when it didn't have the double quotes around it was 1, not 2. I'm confused where you got 2 from. I can only guess that you didn't actually test that exact command, but instead tested just the extract to stdout without piping it:

```powershell
PS [12:18:05] C:\Users\Arnavion\Desktop> Exec cmd /C '"C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y 2>NUL'
cmd /C "C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y 2>NUL exited with code 2
At line:5 char:30
+ [void] ($LASTEXITCODE -and $(throw "$name $arguments exited with code $LASTEXITC ...
+                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (cmd /C "C:\Prog...ted with code 2:String) [], RuntimeException
    + FullyQualifiedErrorId : cmd /C "C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y 2>NUL exited with code 2
```

If you remove the 2>NUL, it's obvious why:

```powershell
PS [12:18:22] C:\Users\Arnavion\Desktop> cmd /C '"C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y'

7-Zip [64] 9.22 beta  Copyright (c) 1999-2011 Igor Pavlov  2011-04-18


Error:
I won't write data and program's messages to same terminal
```

Using 2>NUL does not prevent this error, because 2>NUL will actually be passed to cmd, not to 7z.exe.

Actually, even if you run this directly in cmd.exe, you'll get the same error (I redirected stdout instead of stderr so that the error message is visible):

```batch
[12:21:16] C:\Users\Arnavion\Desktop> "C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y 1>NUL

7-Zip [64] 9.22 beta  Copyright (c) 1999-2011 Igor Pavlov  2011-04-18


Error:
I won't write data and program's messages to same terminal
```

This is a quirk of 7z.exe. The only way to prevent this is to pipe stdout:

```batch
[12:22:33] C:\Users\Arnavion\Desktop> "C:\Program Files\7-Zip\7z.exe" x .\atk-2.14.0.tar.xz -so -y | "C:\Program Files\7-Zip\7z.exe" x -y -si -ttar -o"foo"

<snip large output of successful extraction>

Everything is Ok

Size:       6645760
Compressed: 696064

Everything is Ok

Folders: 14
Files: 482
Size:       6262882
Compressed: 259584
```

**Comment by nacho** [comment]
So I just tried your suggestions but I still get the same error:

gtk3 : ERROR: The parameter is incorrect.
gtk3 :
gtk3 :
gtk3 :
gtk3 : Archives with Errors: 1
gtk3 :
gtk3 : Open Errors: 1
gtk3 :
gtk3 :
gtk3 : Kernel  Time =     2.230 =   17%
gtk3 : User    Time =     0.265 =    2%
gtk3 : Process Time =     2.496 =   19%    Virtual  Memory =      3 MB
gtk3 : Global  Time =    12.916 =  100%    Physical Memory =      5 MB
gtk3 : cmd /c "  "C:\Program Files\7-Zip\7z.exe" x -y -so "C:\mozilla-build\gtk\src\gtk+-3.14.8.tar.xz" | "C:\Program Files\7-Zip\7z.exe" x -y -si -ttar -aoa -o"C:\mozilla-build\gtk\build\x64"" exited with code 2

Also I see that the files were properly extracted, no idea though what the "ERROR: The parameter is incorrect" actually means

**Comment by Arnavion** [comment]
5c99c6c and 9d70913

**Comment by nacho** [comment]
I seem to get these errors now...

cairo : NotSpecified: (:String) [], RemoteException
cairo : 7-Zip [64] 9.36 beta  Copyright (c) 1999-2014 Igor Pavlov  2014-12-26

Processing archive: C:\mozilla-build\gtk\src\cairo-1.14.0.tar.xz

Extracting  cairo-1.14.0.tar

Everything is Ok

Size:       71137280
Compressed: 36584076

Kernel  Time =     0.124 =    3%
User    Time =     1.170 =   35%
Process Time =     1.294 =   39%    Virtual  Memory =     13 MB
Global  Time =     3.307 =  100%    Physical Memory =     14 MB

cairo : cmd /C ""C:\Program Files\7-Zip\7z.exe" x "C:\mozilla-build\gtk\src\cairo-1.14.0.tar.xz" -y -so | "C:\Program Fles\7-Zip\7z.exe" x -o"C:\mozilla-build\gtk\build\x64" -y -si -ttar" exited with code 2

**Comment by Arnavion** [comment]
tomek hit an issue with openssl. Turns out recent versions of 7zip (more recent that mine 9.22) actually support extracting symlinks. 9.22 just creates a text file with the link target, but 9.38 tries to actually create the symlink. The problem is making symlinks requires admin privs on Windows, so this fails with exit code 2 if run without.

Options are that we require admin for building, or we switch back to our own .7z's with the symlink replaced with the actual link target.

@nacho This isn't a problem with cairo though. It extracts fine for me and @tomek with and without admin privs.

**Comment by nacho** [comment]
@Arnavion I guess that makes some sense on the fact it happens to me and not to you or another person that tested the patch. fwiw the 2 extract way seems to work fine for me, what about rolling back to that way?

**Comment by Arnavion** [comment]
The intermediate file wouldn't make any difference if creating symlinks was the problem, so I'm not sure why you're suggesting that.

At any rate we now use our repackaged openssl 7z that replaces the symlinks with the link target so that issue is already fixed.

**Pull request closed without merging**